platform-sdk/docs/proposals/metric-labels/README.md

```diff
@@ -0,0 +1,175 @@
+Labels are key-value pairs that can be used to filter and group metrics.
+Why this proposal uses the name "metric labels" (like prometheus does) other libraries (like micrometer) use the term tags.
+
+## Motivation
+
+Today the metrics module only supports the definition of metrics without adding custom labels.
+While each metric has a  hardcoded category, name, and nodeId no other custom information can be added.
+This proposal suggests adding a new concept called "metric labels" to the metrics module and
+migration the `category`, `name`, and `nodeId` information to labels.
+While the category and name are defined as String values the nodeId is a class called `NodeId`.
+By doing so the metrics module becomes more flexible and can be used 100% independent of the
+platform module since it does not depend on the platform module's `nodeId` anymore.
+
+## Goals
+
+- Add a new concept called "metric labels" to the metrics module
+- Migrate the `category`, `name`, and `nodeId` information to labels
+- Make the metrics module independent of the platform module
+- Provide a public API to define labels for metrics
+- Add labels to the snapshot API
+- The output of the snapshot API should contain the labels
+- The output of the snapshot API must be compatible with the current output (no breaking changes)
+
+## Non-Goals
+
+Even with the new concept the metrics api should be only used in our (hashgraph) code.
+The metrics module should not be used by the end user.
+Here the preferred solution is still that we forward our metrics to a monitoring system that the custom application code uses.
+
+## Design & Architecture
+
+This chapter describes the changes that are needed to add labels to the metrics module.
+
+### Public API
+
+The metrics module will be extended by a new data type called `Label`.
+A label is a key-value pair that can be defined as a `record`:
+
+```java
+record Label(@NonNull String key, @Nullable String value) {}
+``` 
+
+The `Metric` interface will be extended by a new method called `getLabels()` that returns a list of labels:
+
+```java
+interface Metric {
+    
+    //...
+    
+    @NonNull
+    List<Label> getLabels();
+}
+```
+
+> [!IMPORTANT]  
+> I currently ask myself if the method should return `Set`, `Collection` or `List`. Help is appreciated.
+
+Like the name or category, the labels are defined when a metric is created and cannot be changed later.
+Based on that the MetricConfig class will be extended by a new method called `withLabels`:
+
+```java
+abstract class MetricConfig {
+    
+    //...
+    
+    @NonNull
+    MetricConfig withLabels(@NonNull Label... labels) {...}
+    
+    @NonNull
+    MetricConfig withLabels(@NonNull Collection<Label> labels) {...}
+}
+```
+
+By adding labels the unique identifier of a metric is the key and the labels.
+Based on that we need to modify the `Metrics` interface to use the labels as the unique identifier of a metric:
```

Comment:
As far as I understand other metric frameworks, labels are used to count specific observations of a general event,
E.G:  While counting httpRequestsTotal, I would like to classify them into POST, PUT, GET, or by returned status code, or by device generating the request, or by the invoked path.
So each time I observe an event I am tracking with the metric, (assuming a counter) I would use 
httpRequestsTotal.increase(method,deviceType,statusCode, invokedPath).

later I can use promql to get how many get request i received: http_requests_total{method="GET"} or how many requests resulted in an error: http_requests_total{statusCode="500"} 

If I got your idea right, In your proposal, labels are part of the name of the metric, which would make:
a) metric names in the code and in Grafana harder to map.
b) more difficult to query as I would have to know all the labels in advance to create a meaningful query

I also think that category and unit are necessary Prometheus values that one needs to include when mapping our information to Prometheus.

I think that the proposed approach is not compatible with what other frameworks call `labels` or `tags` and including it in the present form in ours would create confusion and eventual incompatibility when we need to map information from existing libraries to swirlds metric framework.

Reply:
Oh, you got me wrong ;) labels **should not** be part of the name. But 1 metric will be identified by its name and its label. Today, a metric is unique only based on its name. With the proposal that will change and you need to give name + labels to get a unique request. Only query for the name can end in a collection of metrics